enhancements/helm3/namespace-scoped-helm-repository-crd.md

```diff
@@ -0,0 +1,200 @@
+  description: my private chart repo
+```
+
+To add a new namespace-scoped Helm repository to a desired namespace:
+```shell
+$ cat <<EOF | oc apply --namespace my-namespace -f -
+apiVersion: helm.openshift.io/v1beta1
+kind: ProjectHelmChartRepository
+metadata:
+  name: stable
+spec:
+  url: https://kubernetes-charts-incubator.storage.googleapis.com
+  displayName: Public Helm stable charts
+  description: Public Helm stable charts hosted on HelmHub
+EOF
+
+$ kubectl get projecthelmchartrepositories --namespace my-namespace
+NAME          AGE
+stable        1m
+```
+
+The addition of namespace-scoped Helm repository does not impact behavior of the existing cluster-scoped Helm repository:
+
+```shell
+$ cat <<EOF | oc apply -f -
+apiVersion: helm.openshift.io/v1beta1
+kind: HelmChartRepository
+metadata:
+  name: stable
+spec:
+  url: https://kubernetes-charts.storage.googleapis.com
+  displayName: Public Helm stable charts
+  description: Public Helm stable charts hosted on HelmHub
+EOF
+
+$ cat <<EOF | oc apply --namespace my-namespace -f -
+apiVersion: helm.openshift.io/v1beta1
+kind: ProjectHelmChartRepository
+metadata:
+  name: incubator
+spec:
+  url: https://kubernetes-charts-incubator.storage.googleapis.com
+  displayName: Public Helm charts in incubator state
+EOF
+
+$ kubectl get helmchartrepositories
+NAME          AGE
+cluster       3h30m
+stable        1m
+
+$ kubectl get projecthelmchartrepositories --namespace my-namespace
+NAME          AGE
+incubator     1m
+```
+
+#### How would the UI render the namespace-scoped Helm repository
+
+As defined in [previous helm enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/helm3/console.md#how-would-the-ui-discover-the-charts), the UI will invoke the `/api/helm/charts/index.yaml` endpoint and the endpoint will proxy requests to the configured chart repositories.
+
+Then the chart repository proxy will use all configured chart repositories, including cluster-scoped and available namespace-scoped Helm repositories in the requested namespace, and deliver to the UI an aggregated index file. The UI then renders the developer catalog with the aggregated index file.
+
+### Implementation Details
+
+[ODC-5713](https://issues.redhat.com/browse/ODC-5713): In the OpenShift Dev Console, add capabilities that allow developers to create Helm chart registries which are namespace scoped, to enable developers to pull in their own Helm charts to their namespace.
+
+![Namespaced Helm Repo Console UI](../helm3/assets/namespace_helm_repo_console_ui.png)
+
+[HELM-244](https://issues.redhat.com/browse/HELM-244): Extend the `/api/helm/charts/index.yaml` API endpoint in OpenShift Dev Console backend to support query parameter `namespace`. For example, a GET request to `/api/helm/charts/index.yaml?namespace=foo` will respond an aggregated `index.yaml` file with entities extracted from both cluster scoped Helm repository and Helm repositories in `foo` namespace.
```

Comment:
so if a user wants to see all the charts in all the repos they have access to, they'd have to keep switching namespaces?  and/or the console would have to make multiple calls to this api(one for each NS the user has access to) and aggregate the results?

Reply:
what permissions are required to invoke `/api/helm/charts/index.yaml` and `/api/helm/charts/index.yaml?namespace=foo` ?

Reply:
1. Yes, like other namespaced CRD, a user needs to switch namespace or pass the `--namespace` flag to access the resource. The usecase for the console API is to display the corresponding helm repositories that user selected in the dev console (the value of dropdown "Project: xxx"). If we need to support displaying multiple namespaces at the same time, we need to return aggregated helm repo in all requested namespaces (this includes the "all NS" case). This enhancement is the first step of supporting namespaced helm repo CRD, which only allows a single namespace in the url parameter.
2. No extra RBAC handling is applied to the `/api/helm/charts/index.yaml` endpoint. So this endpoint authenticate user token just like other API dev console endpoints. However, if the user/token does not have permission to the namespace/CRD, the List CRD request is rejected by k8s go-client calls: https://github.com/openshift/console/pull/10467/files#diff-c8e8620a353bc6becb91e8b2f168e723cdc23d3c59976a93fd5813edbf8990e3R230

Reply:
Don't think the idea is to aggregate all chart repos in all namespaces the user has access to, at least not in this first pass. It is simply the union of repos a cluster scope with all the repos under the current namespace. The feature is scoped to enable members of a project to be able to play with helm charts in the current project without needing admin access to k8 cluster. My assumption is that the user has a role that allows them to have read/list access to ProjectHelmChartRepositoy resource to use the repo and create access to be able to create the repo. A project admin will control that and the helm backend will enforce it.

Reply:
If a user does not have rights to access requested namespace, what will be returned on `/api/helm/charts/index.yaml?namespace=foo` ?

Reply:
@pedjak Only available resouce will be returned, e.g. if the user does not have permission for the `namespace=foo`, then the default cluster-wise Helm repos will be returned, and if no cluster-wise repo is available, an empty list will be returned. Of course we will log the error in the backend but will not explicitly return error to the frontend component. This behavior is captured in the implementation PR: https://github.com/openshift/console/pull/10467.

Reply:
Let's please include the above information in the proposal too.

Reply:
Updated with more details.